packages/core-app-api/src/apis/implementations/auth/microsoft/MicrosoftAuth.ts

```diff
@@ -43,14 +60,108 @@ export default class MicrosoftAuth {
         'email',
         'User.Read',
       ],
+      scopeTransform = scopes => scopes.concat('offline_access'),
     } = options;
 
-    return OAuth2.create({
-      discoveryApi,
-      oauthRequestApi,
-      provider,
-      environment,
-      defaultScopes,
-    });
+    this.environment = environment;
+    this.provider = provider;
+    this.oauthRequestApi = oauthRequestApi;
+    this.discoveryApi = discoveryApi;
+    this.scopeTransform = scopeTransform;
+
+    this.oauth2 = {
+      [MicrosoftAuth.MicrosoftGraphID]: OAuth2.create({
+        discoveryApi: this.discoveryApi,
+        oauthRequestApi: this.oauthRequestApi,
+        provider: this.provider,
+        environment: this.environment,
+        scopeTransform: this.scopeTransform,
+        defaultScopes,
+      }),
+    };
+  }
+
+  private microsoftGraph(): OAuth2 {
+    return this.oauth2[MicrosoftAuth.MicrosoftGraphID];
+  }
+
+  static expectedClaims(scope: string): { aud: string; scp: string } {
+    const scopes = scope.split(' ').map(MicrosoftAuth.parseScope);
+    const firstAudience =
+      scopes.map(({ aud }) => aud).find(aud => aud !== 'openid') ??
+      MicrosoftAuth.MicrosoftGraphID;
+    return {
+      aud: firstAudience,
+      scp: scopes
+        .filter(
+          ({ aud, scp }) =>
+            (aud === 'openid' && scp !== 'offline_access') ||
+            aud === firstAudience,
+        )
+        .map(({ scp }) => scp)
+        .join(' '),
+    };
+  }
+
+  private static parseScope(scope: string): { aud: string; scp: string } {
+    const groups = scope.match(/^(?<resourceURI>.*)\/(?<scp>[^\/]*)$/)?.groups;
+    if (groups) {
+      const { resourceURI, scp } = groups;
+      const aud = resourceURI.replace(/^api:\/\//, '');
+      return { aud, scp };
```

Comment:
Is `scp` short for scope? let's spell that out in that case :grin:, same thing with audience tbh, no need to keep things cryptic here. Same thing for the other occurrences.

Reply:
I was obsessively simulating the exact claims that Azure uses for the sake of my tests -- all gone now